README.md
![sketch-json-color-palette-importer](https://github.com/ziyafenn/sketch-json-color-palette-importer/blob/master/assets/icon.png?raw=true)
# Sketch JSON Color Palette Importer
Create color variables by importing JSON color palettes.

### Sketch >= 69
[Download latest version.](https://github.com/ziyafenn/sketch-json-color-palette-importer/releases/latest)

### Sketch < 69
[Download version 1.1.2](https://github.com/ziyafenn/sketch-json-color-palette-importer/releases/download/v1.1.2/json-color-palette-importer.sketchplugin.zip)

## Description
This plugin will add all the colors from the JSON and create [color variables](https://www.sketch.com/blog/2020/10/15/how-to-make-the-most-of-color-variables/) out of them to be used inside your project.

If you want to create *Shared Styles* instead of *Color Variables*, use version [1.1.2](https://github.com/ziyafenn/sketch-json-color-palette-importer/releases/download/v1.1.2/json-color-palette-importer.sketchplugin.zip).

## Supported JSON Format

```
{
  PALETTE_NAME: {
    COLOR_LABEL: COLOR_HEX,
    ...
  },
  ...
}
```

## Resources to generate colors
You can use [Material Design Palette Generator](https://materialpalettes.com/) to generate accessible color palettes and export them in JSON that will work with this plugin without any modifications.

Another tool generating a JSON file that work out-of-box is [Open Color Tools](http://opencolor.tools/)

You can also use [ColorBox by Lyft Design](https://www.colorbox.io/) to generate a color palette, but it will require slight modification to the JSON code to make it work with the plugin.

## Known issues
Even though generated *color shapes* display that they are linked to the color variables, they are not. Apparently this is an issues with the Sketch API.

## Troubleshooting
If you have any problem with importing JSON code, try to validate it first and then format/prettify it. For validation and formatting you can use any available online tool, such as https://jsonformatter.org/


## TO-DO
- [ ] Check if JSON format is correct
- [ ] Check if any artboard already exisits inside page and create new artboards after them
- [ ] Support any JSON format?


## Licence
MIT


CHANGELOG.md
# Changelog

## [1.2.1]
- Update package description

## [1.2.0]
- Support for Sketch Color Variables
- Better Try/Catch block statement
- Changelog created

## [1.1.2]
- Initial Release

src/my-command.js
/* eslint-disable prefer-destructuring */
// eslint-disable-next-line import/no-unresolved
import sketch from 'sketch'
import {INITIAL_VALUE} from "./shared"

const document = sketch.getSelectedDocument()
const page = document.selectedPage
const Artboard = sketch.Artboard
const Shape = sketch.ShapePath
const UI = sketch.UI
const SharedStyle = sketch.SharedStyle


const newArtboard = (name, y) =>
  new Artboard({
    parent: page,
    name,
    frame: { x: 0, y, width: 1000, height: 100 },
    flowStartPoint: true
  })

const colorShape = (parent, name, x, color) =>
  new Shape({
    parent,
    name,
    frame: { x, y: 0, width: 100, height: 100 },
     style: {
      fills: [color],
    }
  })

  const newSharedStyle = (name, style) => SharedStyle.fromStyle({
    name,
    style: style.style,
    document,
  })

let layerX = 0
let artboardY = 0

export default function() {
  UI.getInputFromUser(
    'Paste color palette JSON',
    {
      initialValue: INITIAL_VALUE,
      numberOfLines: 30,
    },
    (err, value) => {
      if (err) {
        // most likely the user canceled the input
        return
      }
      const parsed = JSON.parse(value)
      for (const palette in parsed) {
        const paletteArboard = newArtboard(palette, artboardY)
        for (const [label, hex] of Object.entries(parsed[palette])) {
         const newColor = colorShape(paletteArboard, label, layerX, hex);
         const newStyle = newSharedStyle(`Color/${palette}/${label}`, newColor)
         newColor.sharedStyleId = newStyle.id
         layerX += 100
        }
        artboardY += 200
        layerX = 0
      }
      
    }
  )
 
}

// TODO fix linter issues
// TODO Check if artboard exists, add after
// TODO Manifest, publish

src/import.js
/* eslint-disable prefer-destructuring */
// eslint-disable-next-line import/no-unresolved
import sketch from 'sketch'
import { INITIAL_VALUE } from './shared'

const document = sketch.getSelectedDocument()
const page = document.selectedPage
const Artboard = sketch.Artboard
const ShapePath = sketch.ShapePath
const UI = sketch.UI
const docSwatches = document.swatches

const newArtboard = (name, y) =>
  new Artboard({
    parent: page,
    name,
    frame: { x: 0, y, width: 1000, height: 100 },
    flowStartPoint: true,
  })

const colorShape = (parent, name, x, color) =>
  new ShapePath({
    parent,
    name,
    frame: { x, y: 0, width: 100, height: 100 },
    style: {
      fills: [{ color }],
      borders: [{ enabled: false }],
    },
  })

const newColorVar = (name, color) =>
  sketch.Swatch.from({
    name,
    color,
  })

let layerX = 0
let artboardY = 0

export default function() {
  UI.getInputFromUser(
    'Paste color palette JSON',
    {
      initialValue: INITIAL_VALUE,
      numberOfLines: 30,
      description:
        'This is an example JSON color palette. Replace it with your own JSON code.',
    },
    (err, value) => {
      if (err) {
        // most likely the user canceled the input
        return
      }

      let parsed
      try {
        parsed = JSON.parse(value)
      } catch (error) {
        UI.alert(
          'Error parsing JSON 😟',
          'Something is wrong with JSON. Please use JSON Validator and Formatter to verify your code.'
        )
      }
      try {
        Object.keys(parsed).forEach(palette => {
          const paletteArboard = newArtboard(palette, artboardY)
          // eslint-disable-next-line no-restricted-syntax
          for (const [label, hex] of Object.entries(parsed[palette])) {
            const colorVarName = `${palette}/${label}`
            const colorVar = newColorVar(colorVarName, hex)
            docSwatches.push(colorVar)
            const colorRef = colorVar.referencingColor
            colorShape(paletteArboard, label, layerX, colorRef)

            layerX += 100
          }
          artboardY += 200
          layerX = 0
        })
        UI.message('🍭 Color Palettes Generated')
      } catch (error) {
        UI.alert(
          'Error creating palette 😟',
          'This is likely a bug in the code of the palette generator, please report.'
        )
      }
    }
  )
}
